app/helpers/chart_helper.rb

```diff
@@ -1,4 +1,40 @@
 module ChartHelper
-  
-  
+  def pluralize(symbol)
+    return symbol.to_s.pluralize.to_sym
+  end
+  def stackedSubtypeChart(grouping)
+    rsubs = Measurement
+              .where("measurements.date >= ? AND measurements.date <= ?",
+                  params[:start].to_date.advance(pluralize(grouping) => -1),
+                  params[:stop].to_date.advance(pluralize(grouping) => 1))
+              .joins(sensor: {subtype: :type})
+              .where("types.resource = ?", @rname)
+              .order('subtypes."usage?"')
+              .group_by_period(grouping, :date).group("subtypes.id, subtypes.name").maximum(:amount)
+    agg = {}
+    rsubs.map do |(date, stype), value|
+      agg[stype] ||= { name: stype, data: [] }
+      agg[stype][:data].append([date,value])

```

Comment:
If you `prepend` or `unshift`, I think you could save yourself the `reverse` on 21.

Reply:
Would just have to reverse again afterwards. Needs to be in chronological order but I need to sub backwards through the array to avoid pre-subtracting:
[5, 7, 8] -> [5, 2, 8] -> [5, 2, 6] instead of
[5, 7, 8] -> [5, 7, 1] -> [5, 2, 1]

Reply:
Might want to look into `bsearch` to get your keys in order the first time. (Ruby hashes have been ordered since... 1.9.3? I think?)
https://stackoverflow.com/a/34115592

Reply:
Oooh, right, of course, the `reverse_each` just changes the order of the iteration, doesn't transform the array itself :derp: